File: src/main/java/com/zunza/buythedip/auth/oauth2/CustomOAuth2FailureHandler.java
package com.zunza.buythedip.auth.oauth2;


import java.io.IOException;
import java.net.URLEncoder;
import java.nio.charset.Charset;

import org.springframework.security.core.AuthenticationException;
import org.springframework.security.web.authentication.SimpleUrlAuthenticationFailureHandler;
import org.springframework.stereotype.Component;

import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;

@Component
@RequiredArgsConstructor
public class CustomOAuth2FailureHandler extends SimpleUrlAuthenticationFailureHandler {

	@Override
	public void onAuthenticationFailure(
		HttpServletRequest request,
		HttpServletResponse response,
		AuthenticationException exception
	) throws IOException, ServletException {
		String errorMessage = URLEncoder.encode(exception.getMessage(), Charset.defaultCharset());
		String redirectUrl = "http://localhost:5173/login/error?message=" + errorMessage;

		response.sendRedirect(redirectUrl);
	}
}

File: src/main/kotlin/com/zunza/buythedip/auth/oauth2/CustomOAuth2FailureHandler.kt
package com.zunza.buythedip.auth.oauth2

import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.security.core.AuthenticationException
import org.springframework.security.web.authentication.SimpleUrlAuthenticationFailureHandler
import org.springframework.stereotype.Component
import java.net.URLEncoder
import java.nio.charset.Charset

@Component
class CustomOAuth2FailureHandler() : SimpleUrlAuthenticationFailureHandler() {
    override fun onAuthenticationFailure(
        request: HttpServletRequest?,
        response: HttpServletResponse?,
        exception: AuthenticationException?
    ) {
        val errorMessage = URLEncoder.encode(exception?.message, Charset.defaultCharset())
        val redirectionUrl = "http://localhost:5173/login/error?message=$errorMessage"
        response?.sendRedirect(redirectionUrl)
    }
}
